pkg/process/metadata/workloadmeta/extractor.go

```diff
@@ -74,11 +85,33 @@ func (w *WorkloadMetaExtractor) Extract(procs map[int32]*procutil.Process) {
 	}
 
 	deadProcs := getDifference(w.cache, newCache)
-	w.grpcListener.writeEvents(deadProcs, newEntities)
 
 	w.cacheMutex.Lock()
 	w.cache = newCache
+	w.cacheVersion++
 	w.cacheMutex.Unlock()
+
+	// Drop previous cache diff if it hasn't been consumed yet
+	select {
+	case <-w.diffChan:
+		// drop message
+		break
+	default:
+	}
+
+	diff := &ProcessCacheDiff{
+		cacheVersion: w.cacheVersion,
+		creation:     newEntities,
+		deletion:     deadProcs,
+	}
+
+	// Do not block on write to prevent Extract caller from hanging e.g. process check
+	select {
+	case w.diffChan <- diff:
+		break
+	default:
+		log.Error("Dropping newer process diff in WorkloadMetaExtractor")
```

Comment:
Won't this get very noisy if we we don't have a consumer. if the core-agent is down or if the core-agent isn't enabled to read the language events?

Reply:
I think it could be a metric

Reply:
That's a really good point. I've dropped this log level to `DEBUG` so it doesn't pollute the log files.

I expect that if language detection is enabled, both gRPC server and client are running but I agree that we can't rely on the fact that the core-agent will be ready to consume events when the Extractor generates the second diff.

Updated in https://github.com/DataDog/datadog-agent/pull/17594/commits/c7463db5298ad98d60286fa9c97d43ba876488a3.

Reply:
In fact I've brought the log level to Error again because this should not happen.
`Extracts` first drops any older diff in the channel before writing to it. So it's expected that all write operations will succeed. If it can't write to the channel, there's an error that we want to capture.